### ISSUE TITLE
Allow priority class values to be updated.

### ISSUE THREAD

**Issue by ravisantoshgudimetla** [kubernetes/kubernetes#99205]
#### What would you like to be added:
I wish to revisit the topic of priority classes being immutable. We made priority classes immutable so that once a priority class is created, we will not be able to update the name or value because of the following reasons:

```
-  May remove config portability: pod specs written for one cluster are no longer guaranteed to work on a different cluster if the same priority classes do not exist in the second cluster.
-  If quota is specified for existing priority classes (at the time of this writing, we don’t have this feature in Kubernetes), adding or deleting priority classes will require reconfiguration of quota allocations.
-  An existing pods may have an integer value of priority that does not reflect the current value of its PriorityClass.
```

While I agree with the first 2 points on changing priority class names, I feel that 3rd point of changing **priority class values** can be re-visited, considering we noticed a lot of people are anyways deleting and re-creating priority classes just to update the value and the updated priority class value is not reflected for existing pods.

#### Why is this needed:

While the deletion and re-creation seems like lesser of both the evils, our experience says that the deletion and re-creation of priority classes is more dangerous and error prone, considering people might delete another priority class, instead of doing just an update where an update can be rejected if there is a name mismatch or value update.

Ref: https://stupefied-goodall-e282f7.netlify.app/contributors/design-proposals/scheduling/pod-priority-api/#modifying-priority-classes

/sig scheduling

**Comment by ravisantoshgudimetla** [comment]
cc @wking @damemi @soltysh

**Comment by wking** [comment]
> -  May remove config portability: pod specs written for one cluster are no longer guaranteed to work on a different cluster if the same priority classes do not exist in the second cluster.

I don't see how in-cluster guards around PriorityClass manipulation can have any bearing on whether Pod specs are portable between clusters.  An admin could set up an admission webhook or RBAC that blocks the removal of a set of well-known PriorityClass names to ensure that Pods consuming those names are portable, but that's not something you can build into a cluster by default outside of [a few, baked-in `systemPriorityClasses` entries][1].

> -  If quota is specified for existing priority classes (at the time of this writing, we don’t have this feature in Kubernetes), adding or deleting priority classes will require reconfiguration of quota allocations.

This sounds like an argument _for_ allowing `value` updates instead of requiring delete/recreate.  Am I missing something?

And my main concern over delete/recreate isn't "might accidentally delete or recreate the wrong PriorityClass" as much as the following flow:

1. Admin deletes PriorityClass A.
2. Pod consuming PriorityClass A is created.
3. Pod blows up, because there is no PriorityClass A.
4. Admin recreates PriorityClass A.
5. Pod schedules successfully on retry, now that PriorityClass A exists again.

By comparison, an in-place `value` update removes the racy gap between 1 and 4, making the change atomic to the updating admin, and as atomic as the server-side wants to make it on the control-plane side.

[1]: https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/pkg/apis/scheduling/v1/helpers.go#L26-L44

**Comment by damemi** [comment]
I agree with Ravi on the 3rd point, allowing updates directly is safer and in my opinion the more expected outcome. In other words, right now if I deleted and recreated a priority class the average user would probably just expect pods using that priority class to use the new value, which is not the case.

**Comment by wking** [comment]
> ...the average user would probably just expect pods using that priority class to use the new value...

I think it's important to manage this expectation via godocs on `value` and other places we doc priority classes.  Because allowing `UPDATE` on PriorityClass's `value` seems orthogonal to addressing priority value updates on existing pods.

**Comment by ingvagabund** [comment]
What will ensure all pods with old value of priority class will get eventually recreated? Keeping some portion of pods with old priority and the rest with a new one might e.g. make descheduler stop/start evicting pods which were/were not supposed to be evicted.

**Comment by alculquicondor** [comment]
What needs to be properly documented is that changing the priority value of a priority class doesn't change the priority of running or pending pods. This has implications for unscheduled pods, but more importantly for eviction by kubelet.
Other than that, I think the request is reasonable. Specially given than a similar effect can be achieved by recreation.

I don't think this requires a KEP, but definitely an API review and a notice to sig node.

**Comment by alculquicondor** [comment]
cc @Huang-Wei

**Comment by Huang-Wei** [comment]
Given that we don't change the semantics that how a recreated/updated PriorityClass impacts existing/upcoming pods, this proposal looks good to me.

Just one concern: for some integrators (not default scheduler), they may rely on PriorityClass deletion to trigger some customized behavior (like recreating/updating pending pods). This proposal may get them impacted.

**Comment by ravisantoshgudimetla** [comment]
> Just one concern: for some integrators (not default scheduler), they may rely on PriorityClass deletion to trigger some customized behavior (like recreating/updating pending pods). This proposal may get them impacted.

I agree and this is concern I  share too. If all of us agree on the above approach, I can send a notice to kubernetes-dev/ & sig-scheduling mailing list and get feedback. We can also go with the approach of having a feature flag and disable  this feature before making it a default choice.

**Comment by alculquicondor** [comment]
Sounds like starting a KEP is the best next step

**Comment by ravisantoshgudimetla** [comment]
I am fine opening a KEP. I hope it is general consensus.

**Comment by alculquicondor** [comment]
cc @ahg-g

**Comment by ravisantoshgudimetla** [comment]
@liggitt - A question for you, would it be ok to `allow updates to objects that were  immutable in previous releases especially if we were achieving the same result by deleting and re-creating the object`

**Comment by liggitt** [comment]
Allowing a previously immutable value to be mutable can be reasonable as long as the things consuming that value are well understood and react to changes correctly. IIUC, about the only in-tree actor that makes use of this value is the admission plugin, which pulls from an informer, so would use the value for future pods, right?

**Comment by ravisantoshgudimetla** [comment]
> IIUC, about the only in-tree actor that makes use of this value is the admission plugin, which pulls from an informer, so would use the value for future pods, right?

That is correct, the priority admission plugin intercepts the pod creation request and translates priority class to priority value. For the pods that are already running, the priority value would still point to old one, say if priority value has been modified or the priority class has been deleted and re-created.

**Comment by alculquicondor** [comment]
It sounds like you could proceed. I would still advice a KEP, because of the readiness questionnaire and to have the change properly documented. Note that this means that you would have to do this next release, as enhancements freeze already happened.

Unless people think a KEP is an overkill for this.

**Comment by ravisantoshgudimetla** [comment]
I think KEP may be an overkill considering we were doing the same with deletion and re-creation pattern but I think it atleast warrants an e-mail to devel community and docs update.

**Comment by soltysh** [comment]
This is changing a stable API, so PRR and a proper documentation is in place, imo.

**Comment by alculquicondor** [comment]
@ravisantoshgudimetla do you plan to work on a KEP for this for 1.23?

**Comment by ravisantoshgudimetla** [comment]
/remove-lifecycle rotten


@alculquicondor - I can work on it. Do folks have review bandwidth? I already have code PR up, I need to put forward KEP.

**Comment by ahg-g** [comment]
Next cycle? Sure.

**Comment by lexisworkingout** [comment]
Hi, for now, is it still not able to use `patch_priority_class()` to update the value of priority class and apply it automatically to running pods?

**Comment by alculquicondor** [comment]
Correct. There are not enough contributors to work on this.

**Comment by wackxu** [comment]
I want to help fix it
/assign

**Comment by XciD** [comment]
Any update on this ?

**Comment by alculquicondor** [comment]
not happening in v1.32

No contributor is looking into it, AFAIK.

**Comment by ravisantoshgudimetla** [comment]
Yeah, if anyone is interested in picking it up, I'd be more than happy to help with reviews and closing this.

**Comment by azr** [comment]
What would be the next steps to get this done ? If I got things correctly:

* make a proposal in [enhancements](https://github.com/kubernetes/enhancements)  ?
* get it approved ?
* repoen the [pr](https://github.com/kubernetes/kubernetes/pull/119458) ?

**Comment by kerthcet** [comment]
Exactly correct.
